zkevm-circuits/src/evm_circuit.rs

```diff
@@ -202,20 +202,23 @@ impl<F: Field> EvmCircuit<F> {
     }
 
     pub fn get_num_rows_required(block: &Block<F>) -> usize {
-        // Start at 1 so we can be sure there is an unused `next` row available
-        let mut num_rows = 1;
-        let evm_rows = block.evm_circuit_pad_to;
+        let evm_rows = block.circuits_params.max_evm_rows;
         if evm_rows == 0 {
-            for transaction in &block.txs {
-                for step in &transaction.steps {
-                    num_rows += step.execution_state.get_step_height();
-                }
-            }
-            num_rows += 1; // EndBlock
+            Self::get_min_num_rows_required(block)
         } else {
-            num_rows += block.evm_circuit_pad_to;
+            block.circuits_params.max_evm_rows + 1
         }
-        num_rows
+    }
+
+    pub fn get_min_num_rows_required(block: &Block<F>) -> usize {
+        let mut num_rows = 0;
+        for transaction in &block.txs {
+            for step in &transaction.steps {
+                num_rows += step.execution_state.get_step_height();
+            }
+        }
+
+        num_rows + 2 // EndBlock and at least one unused
```

Comment:
Not sure this comment is enough to understand why we add `2` precisely. and why `EndBlock` only requires `1`.

Reply:
I really don't actually know the answer, I just refactored the value that had before. An idea of a better explanation for this?

Reply:
Nope.. But we can file an issue for that! :)

Reply:
I don't know the details of the EVM circuit, I guess I should put some time into that. It seems one of the most advanced ones.